src/py21cmfast/_memory.py

```diff
@@ -0,0 +1,706 @@
+    # HIRES_box, HIRES_box_saved
+    num_c_boxes = 2
+
+    """All declared 2LPT boxes (DIM)"""
+    # phi_1 (6 components)
+    if global_params.SECOND_ORDER_LPT_CORRECTIONS:
+        num_c_boxes += 6
+
+    # These are all fftwf complex arrays (thus 2 * size)
+    size_c = (2.0 * (np.float32(1.0).nbytes)) * num_c_boxes * kspace_num_pixels
+
+    return {"python": size_py, "c": size_c}
+
+
+def mem_perturb_field(
+    *,
+    user_params=None,
+):
+    """A function to estimate total memory usage of a perturb_field call."""
+    """Memory usage of Python PerturbedField class."""
+    """All declared HII_DIM boxes"""
+    # density, velocity
+    num_py_boxes_HII_DIM = 2.0
+
+    size_py = num_py_boxes_HII_DIM * (user_params.HII_DIM) ** 3
+
+    # These are all float arrays
+    size_py = (np.float32(1.0).nbytes) * size_py
+
+    """Memory usage within PerturbField.c"""
+    kspace_num_pixels = (float(user_params.DIM) / 2.0 + 1.0) * (user_params.DIM) ** 2
+    hii_kspace_num_pixels = (float(user_params.HII_DIM) / 2.0 + 1.0) * (
+        user_params.HII_DIM
+    ) ** 2
+
+    tot_num_pixels = float(user_params.DIM) ** 3
+    hii_tot_num_pixels = float(user_params.HII_DIM) ** 3
+
+    # LOWRES_density_perturb, LOWRES_density_perturb_saved
+    num_c_boxes_HII_DIM = 2
+
+    # These are all fftwf complex arrays (thus 2 * size)
+    size_c = (
+        (2.0 * (np.float32(1.0).nbytes)) * num_c_boxes_HII_DIM * hii_kspace_num_pixels
+    )
+
+    num_c_boxes_DIM = 0
+    if user_params.PERTURB_ON_HIGH_RES:
+        # HIRES_density_perturb, HIRES_density_perturb_saved
+        num_c_boxes_DIM += 2
+
+        # These are all fftwf complex arrays (thus 2 * size)
+        size_c += (2.0 * (np.float32(1.0).nbytes)) * num_c_boxes_DIM * kspace_num_pixels
+
+        # For resampled_box (double)
+        size_c += (np.float64(1.0).nbytes) * tot_num_pixels
+    else:
+        # For resampled_box (double)
+        size_c += (np.float64(1.0).nbytes) * hii_tot_num_pixels
+
+    return {"python": size_py, "c": size_c}
+
+
+def mem_ionize_box(
+    *,
+    user_params=None,
+    astro_params=None,
+    flag_options=None,
+):
+    """A function to estimate total memory usage of an ionize_box call."""
+    # First do check on inhomogeneous recombinations
+    astro_params = AstroParams(astro_params, INHOMO_RECO=flag_options.INHOMO_RECO)
+
+    # determine number of filtering scales (for USE_MINI_HALOS)
+    if flag_options.USE_MINI_HALOS:
+        n_filtering = (
+            int(
+                np.log(
+                    min(astro_params.R_BUBBLE_MAX, 0.620350491 * user_params.BOX_LEN)
+                    / max(
+                        global_params.R_BUBBLE_MIN,
+                        0.620350491
+                        * user_params.BOX_LEN
+                        / np.float64(user_params.HII_DIM),
+                    )
+                )
+                / np.log(global_params.DELTA_R_HII_FACTOR)
+            )
+        ) + 1
+    else:
```

Comment:
would you reckon we should do the same for minihalo filtering if MINIMIZE_MEMORY?

Reply:
This will also be updated once #220 is merged.